[FSDP] Remove unneeded stream sync from `clip_grad_norm_()`

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#89308 [FSDP] Remove unneeded stream sync from `clip_grad_norm_()`**

We do not need to have the pre-unshard and unshard streams wait for the computation stream because we are not using the pre-unshard or unshard streams in `clip_grad_norm_()`.

The other change is simply avoiding a loop to get `grads`.

Commit:
[FSDP] Remove unneeded stream sync from `clip_grad_norm_()`

[ghstack-poisoned]

Commit:
Update on "[FSDP] Remove unneeded stream sync from `clip_grad_norm_()`"


We do not need to have the pre-unshard and unshard streams wait for the computation stream because we are not using the pre-unshard or unshard streams in `clip_grad_norm_()`.

The other change is simply avoiding a loop to get `grads`.

[ghstack-poisoned]

mrshenli commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1156,32 +1155,34 @@ def clip_grad_norm_(
+                    grads.append(param.grad)
         local_sharded_norm = _get_grad_norm(sharded_params, norm_type).to(
             self.compute_device
         )
```
looks like the `_get_grad_norm` uses grad values. Where do we make sure that reduce scatter on grad is done?

awgu replied:
Post-backward final callback:
https://github.com/pytorch/pytorch/blob/c599cf24ad01a4befa58642f9e703792f53b79fa/torch/distributed/fsdp/_runtime_utils.py#L701-L702

mrshenli reviewed (approved):
stamp to unblock

Commit:
Update on "[FSDP] Remove unneeded stream sync from `clip_grad_norm_()`"


We do not need to have the pre-unshard and unshard streams wait for the computation stream because we are not using the pre-unshard or unshard streams in `clip_grad_norm_()`.

The other change is simply avoiding a loop to get `grads`.

[ghstack-poisoned]

awgu commented:
@pytorchbot merge
